gssapi/WrapToken.go

```diff
@@ -0,0 +1,230 @@
+func getChecksumHeader(flags byte, senderSeqNum uint64) []byte {
+	header := make([]byte, 16)
+	copy(header[0:], []byte{0x05, 0x04, flags, 0xFF, 0x00, 0x00, 0x00, 0x00})
+	ENC.PutUint64(header[8:], senderSeqNum)
+	return header
+}
+
+// Compute the payload + header checksum with the provided key and usage,
+// and compare it to the checksum present in the token
+func (wt *WrapToken) VerifyCheckSum(key types.EncryptionKey, keyUsage uint32) (bool, error) {
+	computed, cErr := wt.ComputeCheckSum(key, keyUsage)
+	if cErr != nil {
+		return false, cErr
+	}
+	if !bytes.Equal(computed, wt.CheckSum) {
+		return false, errors.New(
+			fmt.Sprintf("Checksum mismatch. Computed: %s, Contained in token: %s",
+				hex.EncodeToString(computed), hex.EncodeToString(wt.CheckSum)))
+	}
+	return true, nil
+}
+
+// Parse a wrap token
+// if expectFromAcceptor is true, we expect the token to have been emitted by the gss acceptor,
+// and will check the according flag
+func UnmarshalWrapToken(b []byte, expectFromAcceptor bool) (*WrapToken, error) {
+	// Check if we can read a whole header
+	if len(b) < 16 {
+		return nil, errors.New("bytes shorter than header length.")
+	}
+	// Is the Token ID correct?
+	if !bytes.Equal(GSSWrapTokenID[:], b[0:2]) {
+		return nil, errors.New(
+			fmt.Sprintf("Wrong Token ID. Expected %s, was %s",
+				hex.EncodeToString(GSSWrapTokenID[:]),
+				hex.EncodeToString(b[0:2])))
+	}
+	// Check the acceptor flag
+	flags := b[2]
+	isFromAcceptor := flags&0x01 == 1
+	if isFromAcceptor && !expectFromAcceptor {
+		return nil, errors.New("Unexpected acceptor flag is set. not expecting a token from the acceptor.")
+	}
+	if !isFromAcceptor && expectFromAcceptor {
+		return nil, errors.New("Expected acceptor flag is not set. expecting a token from the acceptor, not the initiator.")
+	}
+	// Check the filler byte
+	if b[3] != FillerByte {
+		return nil, errors.New(
+			fmt.Sprintf("Unexpected filler byte: expecting 0xFF, was %s ", hex.EncodeToString(b[3:4])))
+	}
+	checksumL := ENC.Uint16(b[4:6])
+	// Sanity check on the checksum length
+	if int(checksumL) > len(b)-HdrLen {
+		return nil, errors.New(
+			fmt.Sprintf("Inconsistent checksum length. %d bytes to parse, checksum length is %d", len(b), checksumL))
+	}
+	rrc := ENC.Uint16(b[6:8])
+	seqNum := ENC.Uint64(b[8:16])
+	payload := b[16 : len(b)-int(checksumL)]
+	checksum := b[len(b)-int(checksumL):]
+	return &WrapToken{
+		Flags:    flags,
+		EC:       checksumL,
+		RRC:      rrc,
+		SND_SEQ:  seqNum,
+		Payload:  payload,
+		CheckSum: checksum,
+	}, nil
+}
+
+// Build a new initiator token (acceptor flag will be set to 0) and compute the authenticated checksum.
+// Other flags are set to 0, and the RRC and sequence number are initialized to 0.
+// Note that in certain circumstances you may need to provide a sequence number that has been defined earlier,
+// this is currently not supported.
+func NewInitiatorToken(payload []byte, key types.EncryptionKey) (*WrapToken, error) {
```

Comment:
Exported functions need to be commented for godoc in the format:
```// <Name of function> <description>```
For more information see: https://blog.golang.org/godoc-documenting-go-code